### PR TITLE
Give more detail about eq in book about enums

### PR DESCRIPTION
**Pull request by mdinger**
Second try to address https://github.com/rust-lang/rust/issues/21196 . A lot that was removed at the end basically seemed repetitive showing simple variations on the same type. It seems more effective to just show more variants at the beginning instead.

If you want to pack values into an example, better to use `i32` or some digit than `String` because you don't need the `to_string()` method.

I didn't mention `derive` because:
- I can't explain it (only use it)
- I don't have a link to a good description (maybe rustbyexample but you probably want links internal)
- Giving more detail especially stating that `==` won't work and why should help quite a bit

I didn't `make test` or check links but I will if this will be merged.

@steveklabnik

### CODE REVIEW

File: src/doc/trpl/compound-data-types.md [hunk 1]
```diff
@@ -200,8 +200,62 @@ destructuring `let`, as we discussed previously in 'tuples.' In this case, the
...
+below).
+
+```rust
+enum Character {
+    Digit(i32),
+    Other,
+}
+```
+
+An `enum` variant, can be defined as most normal types. Below are some example
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: you don't need the comma here

File: src/doc/trpl/compound-data-types.md [hunk 2]
```diff
@@ -200,8 +200,62 @@ destructuring `let`, as we discussed previously in 'tuples.' In this case, the
...
+types have been listed which also would be allowed in an `enum`.
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: Hmm, it took me a minute to understand what 'in an enum' meant here, we can workshop this wording, lemme read the rest of this though

File: src/doc/trpl/compound-data-types.md [hunk 3]
```diff
@@ -200,8 +200,62 @@ destructuring `let`, as we discussed previously in 'tuples.' In this case, the
...
+a struct, may or may not hold data. That is, in `Character`, `Digit` is a name
+tied to an `i32` where `Other` is just a name. However, the fact that they are
+distinct makes this very useful.
+
+As with structures, enums don't by default have access to operators such as
+compare ( `==` and `!=`), binary operations (`*` and `+`), and order
+(`<` and `>=`). As such, using the previous `Character` type, the
+following code is invalid:
+
+```{rust, ignore}
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: does this actually ignore with the space here? I thought you couldn't have a space. Maybe that's wrong.

**Comment by steveklabnik** [comment]
Thanks for this @mdinger ! I think that conceptually, I like this patch. We need to do some wordsmithing, but I like the general approach.

File: src/doc/trpl/compound-data-types.md (hunk 3 above)
Comment (mdinger) [inline src/doc/trpl/compound-data-types.md]: Idk, I'll change it.

**Comment by mdinger** [comment]
Oh good. Thanks for the quick look over. Long lead times get frustrating.

Actually, I think I may be able to work in `#[derive(...)]` but I gotta think about it. The way it's usually used, it tends to feel like *black magic*. Flip the switch and everything works but we don't explain how.

File: src/doc/trpl/compound-data-types.md (hunk 2 above)
Comment (mdinger) [inline src/doc/trpl/compound-data-types.md]: Definitely.

**Comment by steveklabnik** [comment]
Yeah, I'm at an airport right now, about to fly from Europe to Australia, so apologies if there ends up being a longer time for reviewing. For the benefit of other committers, since this is mostly grammar nitpicking, I'm comfortable r+ing this, since I'm going to end up doing a full editing pass later, and it is just nitpicks.

> Actually, I think I may be able to work in #[derive(...)] 

I don't think that's a good idea: we haven't covered _any_ of the relevant information yet.

File: src/doc/trpl/compound-data-types.md (hunk 1 above)
Comment (mdinger) [inline src/doc/trpl/compound-data-types.md]: fixed

**Comment by mdinger** [comment]
Okay. I'll do make check and check the links in the morning.

I do tend to agree but this still leaves it open to the reader deciding "I **want** to define `Ordering` myself". But it'll just have to be dealt with, at least for now.

By the way, you guys typically have amazing response times. Especially when it's just a few minutes and since you deal with 100's of emails a day.

**Comment by steveklabnik** [comment]
> By the way, you guys typically have amazing response times. Especially when it's just a few minutes and since you deal with 100's of emails a day.

Thanks. :smile:

**Comment by mdinger** [comment]
`make check` passed and the links work. I don't know what this formatting note means but it didn't error. The formatting section looks different now. Weird.
```
check: formatting
...
249 error codes
highest error code: E0315
```

@steveklabnik r?

**Pull request merged**